### DISCUSSION TITLE
Connecting to custom tile server

### DISCUSSION THREAD

**Discussion by domagojhack** [raphaelquast/EOmaps#105] [Q&A] [answered]
Me again, I keep getting into the intricacies of EOmaps. Here I am asking about connecting to custom tile servers.

Actually, I managed to set up a private server that serves OpenStreetMap data for my application. I only see instructions to connect to WMS-type providers using GetCapabilites.xml links in docs. My server serves tiles in the following matter:

`
http://myserver.serv/hot/{x}/{y}/{z}.png
`

In the source, (`_webmap.py`) I see there are some classes referring to XYZ data and maybe could be used to load up such data but I can not wrap my head around how to access them from the `m` EOmaps object. How can I do this?

**Comment by raphaelquast** [comment, marked as answer]
Hey, 
good question... I never tried to use a local server but in theory it should work just as any other server...
(if you have a nice overview on how to setup local tileservers, I'd love to get a look!)

Basically you should be able to use `m.add_wms.get_service` to get a WebMap object for your url and then use it just as any other WebMap service.
(check the docs on [Using custom webmap services](https://eomaps.readthedocs.io/en/latest/api.html#using-custom-webmap-services) and [m.add_wms.get_service](https://eomaps.readthedocs.io/en/latest/generated/eomaps._containers.wms_container.get_service.html) )

If all goes well, that should work:
```python
m = Maps(Maps.CRS.GOOGLE_MERCATOR)
service = m.add_wms.get_service("http://myserver.serv/hot/{x}/{y}/{z}.png", service_type="xyz")
service.add_layer.xyz_layer()
```

**Reply by domagojhack** [reply]
Thanks for the reply. Yeah, I tried already like this but got some mixed results. Now I realized that my server link is actually different.
Instead of `http://myserver.serv/hot/{x}/{y}/{z}.png` it is `http://myserver.serv/hot/{z}/{x}/{y}.png`

Full minimal working example is:

```python
from eomaps import Maps

m = Maps()

service = m.add_wms.get_service(
    "http://yourserver.server/hot/{z}/{x}/{y}.png",
    service_type="xyz")

service.add_layer.xyz_layer()

m.show()
```
![image](https://user-images.githubusercontent.com/2804094/180148085-ecd7583b-3bc0-4e83-bfb3-1d270640efee.png)


To set up my private OSM server I used the [switch2osm](https://switch2osm.org/serving-tiles/) framework. I had to do a bit of tinkering but it is doable in one day.

**Reply by raphaelquast** [reply]
Great! Nice to see that things are working properly for offline OSM-tileservers!   (I'll definitely have to checkout swith2osm as well 😅 )
